execute/outcome.go

```diff
@@ -121,31 +121,59 @@ func (p *Plugin) getMessagesOutcome(
 	observation exectypes.Observation,
 	previousOutcome exectypes.Outcome,
 ) exectypes.Outcome {
-	commitReports := previousOutcome.PendingCommitReports
+	commitReports := make([]exectypes.CommitData, 0)
 	costlyMessagesSet := mapset.NewSet[cciptypes.Bytes32]()
 	for _, msgID := range observation.CostlyMessages {
 		costlyMessagesSet.Add(msgID)
 	}
 
+	// First ensure that all observed messages has hashes and token data.
+	if err := validateHashesExist(observation.Messages, observation.Hashes); err != nil {
+		p.lggr.Errorw("validate hashes exist: %w", err)
+		return exectypes.Outcome{}
+	}
+	if err := validateTokenDataObservations(observation.Messages, observation.TokenData); err != nil {
+		p.lggr.Errorw("validate token data observations: %w", err)
+		return exectypes.Outcome{}
+	}
+
+	reports := observation.CommitReports.Flatten()
 	// add messages to their commitReports.
-	for i, report := range commitReports {
+	for i, report := range reports {
 		report.Messages = nil
 		report.CostlyMessages = nil
 		for j := report.SequenceNumberRange.Start(); j <= report.SequenceNumberRange.End(); j++ {
 			if msg, ok := observation.Messages[report.SourceChain][j]; ok {
+				// Always add the message to the report, even if it is pseudo-deleted.
 				report.Messages = append(report.Messages, msg)
+
+				report.Hashes = append(report.Hashes, observation.Hashes[report.SourceChain][j])
+
+				// Add the pseudo-deleted status to the report.
+				if observation.PseudoDeleted[report.SourceChain] != nil {
+					report.MessagePseudoDeleted = append(report.MessagePseudoDeleted,
+						observation.PseudoDeleted[report.SourceChain][j],
+					)
+				} else {
+					report.MessagePseudoDeleted = append(report.MessagePseudoDeleted, false)
+				}
 				if costlyMessagesSet.Contains(msg.Header.MessageID) {
 					report.CostlyMessages = append(report.CostlyMessages, msg.Header.MessageID)
 				}
```

Comment:
Can you remove these now?

Reply:
Yes. I added verification code [earlier](https://github.com/smartcontractkit/chainlink-ccip/pull/348/files#diff-da84b98cf1e979eb462000a946f90ba56747f7af3453826ddc0280bdd2186df5R131-R138)